### PR TITLE
No parenthesis for Flow shorthand with one arg

### PR DESCRIPTION
**Pull request by yamafaktory**
Experimental PR trying to solve #969.

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -2062,8 +2062,12 @@ function printFunctionParams(path, print, options) {
     return concat(["(", join(", ", printed), ")"]);
   }
 
+  const isFlowShorthandWithOneArg = fun.type === "FunctionTypeAnnotation" &&
+    parent.type === "ObjectTypeProperty" && fun.params.length === 1 &&
```
Comment (yamafaktory) [inline src/printer.js]: Oh, I don't need that...

**Comment by vjeux** [comment]
Can you add test cases for:

```js
type T = { method: (a) => void };
 type T = { method(a): void };
 declare class X { method(a): void }
 declare function f(a): void;
 var f: (a) => void;
```

**Comment by vjeux** [comment]
Yeah, that's what I suspected, the "normal" solution completely breaks:

```js
type T = { method: a => void };
type T = { methoda: void };
declare class X { methoda: void }
declare function fa: void;
var f: a => void;
```

Also, this is just a sample of all the ways you can combine this ast. I'm honestly super scared about breaking something here.

**Comment by yamafaktory** [comment]
Yeah I've just saw that... will try to see what I can do...

**Comment by yamafaktory** [comment]
Checking the equality of the `start` of the `fun` and the one of the first param seems to fix it.

**Comment by yamafaktory** [comment]
Ok, this is now a regression for some other tests.

**Comment by yamafaktory** [comment]
@vjeux I think I got something better 🤞. At least the tests are passing, but I maybe need more test cases, not quite sure about it.

**Comment by yamafaktory** [comment]
@cpojer do you have some other examples coming in mind so that I can test my implementation against it?

File: tests/flow/annot/annot.js [hunk 2]
```diff
@@ -58,3 +58,17 @@ var zer : null = null;
...
+
+type T = { method: (a) => void };
+
+type T = { method(a): void };
+
+declare class X { method(a): void }
+
+declare function f(a): void;
+
+var f: (a) => void;
```
Comment (vjeux) [inline tests/flow/annot/annot.js]: A few more things to test:

```js
interface F { m(string): number };
interface F { m: (string) => number };
function f(o: { f: (string) => void }) {}
function f(o: { f(string): void }) {}
```

Comment (yamafaktory) [inline tests/flow/annot/annot.js]: Done and passing with my implementation!

File: src/printer.js [hunk 3]
```diff
@@ -2062,8 +2062,12 @@ function printFunctionParams(path, print, options) {
     return concat(["(", join(", ", printed), ")"]);
   }
 
+  const isFlowShorthandWithOneArg = fun.type === "FunctionTypeAnnotation" &&
```
Comment (yamafaktory) [inline src/printer.js]: Ok, done. But I needed to create a similar one called `isTypeAnnotationAFunction` and also keep a checking for TypeAlias too 😄.

File: src/printer.js [hunk 4]
```diff
@@ -3022,6 +3026,16 @@ function isObjectTypePropertyAFunction(node) {
     util.locStart(node.key) !== util.locStart(node.value);
 }
 
+// Hack to differentiate between the following two which have the same ast
+// declare function f(a): void;
+// var f: (a) => void;
+function isTypeAnnotationAFunction(node) {
```
Comment (yamafaktory) [inline src/printer.js]: Similar to `isObjectTypePropertyAFunction`.

**Comment by yamafaktory** [comment]
@vjeux I've just noticed the build passed for v4 only and failed for no reason of the other two.

File: src/printer.js [hunk 5]
```diff
@@ -2062,8 +2062,12 @@ function printFunctionParams(path, print, options) {
     return concat(["(", join(", ", printed), ")"]);
   }
 
+  const isFlowShorthandWithOneArg = (isObjectTypePropertyAFunction(parent) ||
+    isTypeAnnotationAFunction(parent) || parent.type === "TypeAlias") &&
+    fun.params.length === 1 && fun.params[0].name === null && fun.rest === null;
```
Comment (vjeux) [inline src/printer.js]: Looks like this has less cases than arrow function:

```js
        n.params.length === 1 &&
        !n.rest &&
        n.params[0].type === "Identifier" &&
        !n.params[0].typeAnnotation &&
        !n.params[0].leadingComments &&
        !n.params[0].trailingComments &&
        !n.params[0].optional &&
        !n.predicate &&
        !n.returnType
```

https://github.com/prettier/prettier/blob/master/src/printer.js#L304-L313

Can you add the following tests as well:

```js
type f = (...arg) => void;
type f = (/* comment */ arg) => void;
type f = (arg /* comment */) => void;
type f = (?arg) => void;
```

**Comment by vjeux** [comment]
Looks like both builds got stuck downloading node after 50min. I just restarted them.

File: src/printer.js (hunk 5 above)
Comment (yamafaktory) [inline src/printer.js]: Ok, I will do it!

**Comment by yamafaktory** [comment]
@vjeux done!

**Comment by vjeux** [comment]
Boom! Thanks a lot!

**Pull request merged**